File: 22_专题/前缀与差分/1371. 每个元音包含偶数次的最长子字符串.ts
/**
 * @param {string} s
 * @return {number}
 * @summary 滑动窗口（这里是可变滑动窗口）我们需要扩张和收缩窗口大小，而这里不那么容易
 * 关注五个数的状态:每个数的状态为0/1 状态压缩即可
 */
const findTheLongestSubstring = function (s: string): number {
  let res = 0
  let state = 0b00000
  // 每种状态最早出现的索引
  const visited = new Map<number, number>([[state, -1]])
  const record = {
    a: 1 << 0,
    e: 1 << 1,
    i: 1 << 2,
    o: 1 << 3,
    u: 1 << 4,
  } as { [k: string]: number }

  for (let i = 0; i < s.length; i++) {
    if (s[i] in record) state ^= record[s[i]]
    if (!visited.has(state)) visited.set(state, i)
    else res = Math.max(res, i - visited.get(state)!)
  }

  return res
}

console.log(findTheLongestSubstring('eleetminicoworoep'))

export {}

File: 22_专题/前缀与差分/经典题/状态压缩+前缀/1371. 每个元音包含偶数次的最长子字符串.py
"""每个元音包含偶数次的最长子字符串"""

VOWEL = set("aeiou")


class Solution:
    def findTheLongestSubstring(self, s: str) -> int:
        res, state = 0, 0
        preState = dict({0: -1})
        for i, char in enumerate(s):
            if char in VOWEL:
                state ^= 1 << (ord(char) - 97)
            if state not in preState:
                preState[state] = i
            else:
                res = max(res, i - preState[state])
        return res
